realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java

```diff
@@ -528,20 +529,26 @@ private void emitCopyMethod(JavaWriter writer) throws IOException {
                 className, // Return type
                 "copy", // Method name
                 EnumSet.of(Modifier.PUBLIC, Modifier.STATIC), // Modifiers
-                "Realm", "realm", className, "newObject", "boolean", "update"); // Argument type & argument name
+                "Realm", "realm", className, "newObject", "boolean", "update", "Map<Object, RealmObject>", "cache"); // Argument type & argument name
 
         writer.emitStatement("%s realmObject = realm.createObject(%s.class)", className, className);
+        writer.emitStatement("cache.put(newObject, realmObject)");
         for (VariableElement field : fields) {
             String fieldName = field.getSimpleName().toString();
+            String fieldType = field.asType().toString();
             if (typeUtils.isAssignable(field.asType(), realmObject)) {
                 writer
                     .emitEmptyLine()
-                    .emitStatement("%s %s = newObject.%s()", Utils.getFieldTypeSimpleName(field), fieldName, getters.get(fieldName))
+                    .emitStatement("%s %s = newObject.%s()", fieldType, fieldName, getters.get(fieldName))
                     .beginControlFlow("if (%s != null)", fieldName)
-                        .emitStatement("realmObject.%s(%s.copyOrUpdate(realm, %s, update))",
-                                setters.get(fieldName),
-                                Utils.getProxyClassSimpleName(field),
-                                fieldName)
+                        .beginControlFlow("if (cache.containsKey(%s))", fieldName)
```

Comment:
cache.containsKey() + cache.get() is slow.
Use cache.get() and check for null instead